source/develop/release-management/features/storage/multipath-events.html.md

```diff
@@ -0,0 +1,193 @@
+
+# Multipath Events
+
+
+## Overview
+
+This feature adds the ability to display events to the user in
+the engine UI when multipath devices paths change their status on a host.
+The supported events are:
+- Faulty multipath paths on host "HOST_NAME" on devices: "GUID", "GUID" ... 
+- No faulty multipath paths on host "HOST_NAME"
+The purpose of this feature is to help users troubleshoot storage issues
+before they become critical when a multipath device is lost, and to
+make it easier to troubleshoot cases when some LUNs are not available
+because all paths became faulty.
+
+
+## Owner
+
+- Name: [Fred Rolland](https://github.com/rollandf)
+- Email: <frolland@redhat.com>
+
+
+## Detailed Description
+
+
+### How it works
+
+When a host is connected to a DC (Storage Pool), Vdsm will register
+to udev multipath events. Vdsm will also build a map of all the faulty paths. 
+For each event, Vdsm will update the list, according to the event.
+The details of the faulty paths will be part of the statistics provided by
+'Host.getStats' verb.
+Engine polls the hosts for the statistics every 10 seconds.
+Engine creates an audit log event according to the information provided
+in the statistics. Only if there are changes in the faulty paths report,
+an event will be created.
+Once the host is disconnected from the DC, Vdsm will unregister from
+the 'udev' events.
+
+
+### New vdsm API
+
+New property in 'HostStats': 'faultyPathsMap' which is a mapping of faulty paths
+indexed by the device mapper GUID.
+
+
+#### Faulty Paths Map
+
+The key of the map is the device mapper GUID.
+The value is a list of faulty device mapper paths.
+
+### Vdsm multipath event listener
+
+Introduce a new Vdsm internal service for listening to multipath events:
+
+- Code interested in receiving multipath events will subscribe a callback to receive the event
+- When an event is received, all subscribers will receive the events (in another thread).
+- If nobody is registered for events, the events are dropped
+- The multipath event monitor will be started when connecting to the pool
+- The multipath event monitor will be stopped when disconnecting from the pool
+
+
+A Udev event monitor is introduced, listening to device-mapper udev events.
+The relevant events for this feature are PATH_FAILED and PATH_REINSTATED:
+
+Variable Name: DM_ACTION
+Uevent Action(s): KOBJ_CHANGE
+Type: string
+Description:
+Value: Device-mapper specific action that caused the uevent action.
+	PATH_FAILED - A path has failed.
+	PATH_REINSTATED - A path has been reinstated.
+
+- The user can disable this feature in a specific host in Vdsm configuration file
+
+In the future, the monitor can be used for other code, for example maintaining LVM filter.
+
+### Vdsm multipath statistics handler
+
+To report multipath statistics to engine, we will add a multipath statistics handler.
+
+- The handler builds a map of all the faulty paths during startup.
+- The handler subscribes to multipath events during startup
```

Comment:
We must first register for events - only then we can check the current status.

We can process events only when we finished to fetch the current status, otherwise we may loose events.

Reply:
done